lms/djangoapps/grades/models.py

```diff
@@ -201,44 +203,55 @@ def bulk_create(cls, course_key, block_record_lists):
             )
             for brl in block_record_lists
         ])
-        cls._update_cache(course_key, created)
+        cls._update_cache(user_id, course_key, created)
         return created
 
     @classmethod
-    def bulk_get_or_create(cls, block_record_lists, course_key):
+    def bulk_get_or_create(cls, user_id, course_key, block_record_lists):
         """
         Bulk creates VisibleBlocks for the given iterator of
-        BlockRecordList objects for the given course_key, but
+        BlockRecordList objects for the given user and course_key, but
         only for those that aren't already created.
         """
-        existent_records = cls.bulk_read(course_key)
-        non_existent_brls = {brl for brl in block_record_lists if brl.hash_value not in existent_records}
-        cls.bulk_create(course_key, non_existent_brls)
+        cached_records = cls.bulk_read(user_id, course_key)
```

Comment:
Are these records always previously cached?  That is, if `_initialize_cache` was just called, then we don't need to check for `already_created` again, right?

Reply:
My question was more about whether most code paths through this usually have a prior call to prefetch the data.  When they don't prefetch, then the extra SQL call to check for changes is unnecessary since the cache wasn't previously populated.

Reply:
Ah, that's probably what's happening.  Maybe we can do that instead and eliminate this extra SQL query.